### PR TITLE
feat(extension/googleclientauth): support id-token

### PR DESCRIPTION
**Pull request by rinx**
This PR adds support for generating [id-token](https://cloud.google.com/docs/authentication/token-types#id).

I'm planning to deploy OpenTelemetry collector as a Cloud Run application.
I'd like to use [Cloud Run's service-to-service authentication](https://cloud.google.com/run/docs/authenticating/service-to-service) or [Cloud IAP's programmatic authentication](https://cloud.google.com/iap/docs/authentication-howto#obtaining_an_oidc_token_for_the_default_service_account) to keep it private.

By supporting id-token in the googleclientauth extension, OpenTelemetry collector can be used as a client for the private collector deployed as a Cloud Run.

### CODE REVIEW

**Comment by dashpole** [comment]
> I'm planning to deploy OpenTelemetry collector as a Cloud Run application.

To clarify, you are planning to run the OTel collector as a service in cloud run, and then use a different collector running elsewhere to send to the cloud run service?

File: extension/googleclientauthextension/factory.go [hunk 1]
```diff
@@ -54,10 +56,26 @@ func (ca *clientAuthenticator) Start(ctx context.Context, _ component.Host) erro
...
 }
 
+func (ca *clientAuthenticator) newTokenSource(ctx context.Context, creds *google.Credentials) (oauth2.TokenSource, error) {
+	switch ca.config.TokenFormat {
+	case "id_token":
+		if ca.config.Audience == "" {
+			return nil, errors.New("audience must be specified")
+		}
+		return idtoken.NewTokenSource(ctx, ca.config.Audience)
+	default:
```
Comment (damemi) [inline extension/googleclientauthextension/factory.go:74]: we should also update `(cfg *Config) Validate()` to check that the input is one of the valid values and return an error if not

**Review by damemi — CHANGES_REQUESTED** [review]

**Comment by rinx** [comment]
> To clarify, you are planning to run the OTel collector as a service in cloud run, and then use a different collector running elsewhere to send to the cloud run service?

That's right.
I'd like to run an OTel collector as a receiver service in Cloud Run, and deploy "sender" OTel collectors somewhere else (e.g. Cloud Run, GKE, or AWS services via workload identity federation).
This PR is for the "sender" role collectors.

File: extension/googleclientauthextension/factory.go (hunk 1 above)
Comment (rinx) [reply extension/googleclientauthextension/factory.go:74]: I added validation for `token_format`. https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/851/commits/d50d02bdd1648891cf82f8a7cbe86ae0489e54fe

File: extension/googleclientauthextension/config.go [hunk 2]
```diff
@@ -36,13 +57,24 @@ type Config struct {
 	// See https://datatracker.ietf.org/doc/html/rfc6749#section-3.3
 	Scopes []string `mapstructure:"scopes,omitempty"`
 
+	// TokenFormat specifies which type of token will be generated.
+	// default: access_token
+	TokenFormat string `mapstructure:"token_format,omitempty"`
```
Comment (dashpole) [inline extension/googleclientauthextension/config.go:62]: Nit: should this be token_type? https://cloud.google.com/docs/authentication/token-types

File: extension/googleclientauthextension/config.go [hunk 3]
```diff
@@ -15,9 +15,30 @@
...
 
+type TokenFormat uint8
+
+const (
+	AccessToken TokenFormat = 1 + iota
+	IDToken
+)
+
+// toTokenFormat returns TokenFormat
+func toTokenFormat(t string) (TokenFormat, error) {
```
[the remark this answers is not in the document]
Comment (damemi) [reply extension/googleclientauthextension/config.go:32]: We should probably at least define the strings as constants right?

File: extension/googleclientauthextension/factory.go [hunk 4]
```diff
@@ -54,10 +56,37 @@ func (ca *clientAuthenticator) Start(ctx context.Context, _ component.Host) erro
...
+	case IDToken:
+		if ca.config.Audience == "" {
+			return nil, errors.New("audience must be specified")
+		}
+		return idtoken.NewTokenSource(
+			ctx,
+			ca.config.Audience,
+			idtoken.WithCredentialsJSON(creds.JSON),
+		)
+	case AccessToken:
```
[the remark this answers is not in the document]
Comment (rinx) [reply extension/googleclientauthextension/factory.go]: I removed the redundant branch.

File: extension/googleclientauthextension/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rinx) [reply extension/googleclientauthextension/config.go]: I removed the enum and added the constants.

File: extension/googleclientauthextension/config.go (hunk 2 above)
Comment (rinx) [reply extension/googleclientauthextension/config.go]: renamed. :+1:

**Review by dashpole — COMMENTED** [review]
can you update http_test and grpc_test with cases that use the id_token token type, and check the resulting headers?

**Comment by rinx** [comment]
> can you update http_test and grpc_test with cases that use the id_token token type, and check the resulting headers?

added test cases that uses `id_token`.
However, I couldn't write a case for checking headers, because `google.golang.org/api/idtoken` requires a real GCP credentials.json for generating a token.

I've already confirm that this code works correctly in my real GCP project.

**Review by dashpole — APPROVED** [review]

**Comment by rinx** [comment]
@damemi Could you please review this PR again?

**Review by damemi — APPROVED** [review]

**Pull request merged**